pkg/lib/interaction/nodes/authn_end.go

```diff
@@ -30,12 +31,28 @@ type EdgeAuthenticationEnd struct {
 }
 
 func (e *EdgeAuthenticationEnd) Instantiate(ctx *interaction.Context, graph *interaction.Graph, input interface{}) (interaction.Node, error) {
-	return &NodeAuthenticationEnd{
+	node := &NodeAuthenticationEnd{
 		Stage:                 e.Stage,
 		AuthenticationType:    e.AuthenticationType,
 		VerifiedAuthenticator: e.VerifiedAuthenticator,
 		RecoveryCode:          e.RecoveryCode,
-	}, nil
+	}
+
+	if err := node.IsFailure(); err != nil {
+		userID := graph.MustGetUserID()
+		user, err := ctx.Users.Get(userID)
+		if err != nil {
+			return nil, err
+		}
+		err = ctx.Events.DispatchEvent(&nonblocking.UserFailedAuthenticationEventPayload{
+			User: *user,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Thinking should we also dispatch `UserFailedAuthenticationEventPayload` if the user failed to select identity? e.g. user input invalid email. It should be useful for audit purpose, e.g. one IP keep trying to burst force the email address.
I tried auth0, they had an audit event for that case, the event context will have the `username` that the user is trying and an empty user id.

What do you think?

Reply:
But in that case, we cannot associate the event with any user. The event can only be viewed in the audit log screen. Should we add a new event type for that? So the admin can select this particular event type to view all malicious brute force attack.

Reply:
Sure, any ideas for the names? We need to have a name that is easy to identify 🤔 For auth0, they are `Failed Login (invalid email/username)` and `Failed Login (wrong password)`